### PR TITLE
update and improve json schemas to draft 7

### PR DESCRIPTION
**Pull request by dsifford**
Follow up from #153, closes https://github.com/citation-style-language/schema/issues/127.

This _should_ be backwards compatible with the old schemas. And because of that, there's some glaring issues with some of the fields and types scatter throughout that were grandfathered in. These I think should be discussed and improved upon in another PR.

If you'd like, copy and paste the `csl-data.json` schema to [this validator](https://www.jsonschemavalidator.net/) to give it a shot. The `csl-citation.json` schema will not validate because it references the current broken schema url.

Let me know if you have any questions.

Ping: @rmzelle @dhimmel

### CODE REVIEW

**Comment by rmzelle** [comment]
Great, thanks. Will you address the merge conflict?

And can you briefly explain how you managed to upgrade the schemas? Are there software libraries to upgrade schemas to newer versions, or was it hand-work?

> This _should_ be backwards compatible with the old schemas. And because of that, there's some glaring issues with some of the fields and types scatter throughout that were grandfathered in. These I think should be discussed and improved upon in another PR.

Can you give an example of what I should be looking for in this PR?

**Comment by dsifford** [comment]
Whoops, didn't see that there was a conflict.

> And can you briefly explain how you managed to upgrade the schemas? Are there software libraries to upgrade schemas to newer versions, or was it hand-work?

Mostly by hand.. I used [quicktype](https://quicktype.io/app) to quickly generate raw object definitions, but most of those had to be adjusted for better accuracy. Then I used the validator I liked above to double check that the object types checked out using a few known-accurate objects.

I also used a few schemas on [schemastore.org](http://schemastore.org/json/) for inspiration.

> Can you give an example of what I should be looking for in this PR?

I'll annotate some lines where changes were made and where I think the schema can be improved in this PR after I fix the conflict.

File: csl-citation.json [hunk 1]
```diff
@@ -1,101 +1,103 @@
...
+                        "string",
+                        "number",
+                        "boolean"
+                    ]
+                },
+                "uris": {
+                    "type": "array",
+                    "items": {
+                        "type": "string",
+                        "format": "uri"
```
Comment (dsifford) [inline csl-citation.json]: Changed this to specifically require `uri` format.

File: csl-citation.json [hunk 2]
```diff
@@ -1,101 +1,103 @@
...
+        "citationItemDefinition": {
+            "type": "object",
+            "properties": {
+                "id": {
+                    "type": [
+                        "string",
+                        "number"
+                    ]
+                },
+                "itemData": {
```
Comment (dsifford) [inline csl-citation.json]: This is not used whatsoever in citeproc-js.

File: csl-citation.json [hunk 3]
```diff
@@ -1,101 +1,103 @@
...
+                        "format": "uri"
+                    }
+                }
+            },
+            "additionalProperties": false
+        }
+    },
     "type": "object",
     "properties": {
         "schema": {
```
Comment (dsifford) [inline csl-citation.json]: Why is this necessary?

File: csl-citation.json [hunk 4]
```diff
@@ -1,101 +1,103 @@
...
             "type": "string",
-            "required": true,
-            "enum" : [
-                "https://github.com/citation-style-language/schema/raw/master/csl-citation.json" 
-            ] 
+            "format": "uri"
         },
         "citationID": {
```
Comment (dsifford) [inline csl-citation.json]: Wishlist: This should be called `id`.

File: csl-citation.json [hunk 5]
```diff
@@ -1,101 +1,103 @@
...
+                    "type": "number"
+                }
             },
-            "additionalProperties" : false 
-        } 
+            "additionalProperties": false
+        }
     },
-    "additionalProperties" : false 
+    "required": ["schema", "citationID", "citationItems"],
```
Comment (dsifford) [inline csl-citation.json]: Changed this to include `citationItems` because it should always return an array, even if empty IMHO.

File: csl-data.json [hunk 6]
```diff
@@ -1,431 +1,440 @@
...
+                            "properties": {
+                                "date-parts": {
+                                    "type": "array",
+                                    "items": {
+                                        "type": "array",
+                                        "items": {
+                                            "type": "number"
+                                        },
+                                        "maxItems": 3,
+                                        "minItems": 1
```
Comment (dsifford) [inline csl-data.json]: Added `minItems`

File: csl-data.json [hunk 7]
```diff
@@ -1,431 +1,440 @@
...
-                                    "type": "string" 
-                                },
-                                "parse-names" : {
-                                    "type": [
-                                        "string",
-                                        "number",
-                                        "boolean" 
-                                    ] 
-                                } 
+                                "raw": {
```
Comment (dsifford) [inline csl-data.json]: Made this an "either-or" between `raw`, and `date-parts`.

File: csl-data.json [hunk 8]
```diff
@@ -1,431 +1,440 @@
...
-                } 
-            },
-            "accessed": {
-                "id": "date-variable",
-                "type": [
+                    ]
+                },
+                {
+                    "properties": {
+                        "circa": {
```
Comment (dsifford) [inline csl-data.json]: Boolean? What is this exactly.

File: csl-data.json [hunk 9]
```diff
@@ -1,431 +1,440 @@
...
+                                "string",
+                                "number",
+                                "boolean"
+                            ]
+                        },
+                        "season": {
+                            "type": [
+                                "number"
+                            ],
+                            "minimum": 1,
```
Comment (dsifford) [inline csl-data.json]: Changed this to be a better spec fit.

File: csl-data.json [hunk 10]
```diff
@@ -1,431 +1,440 @@
...
-                                    "items": {
-                                        "type": [
-                                            "string",
-                                            "number" 
-                                        ] 
-                                    },
-                                    "maxItems": 3 
-                                },
-                                "maxItems": 2 
+                            "literal": {
```
Comment (dsifford) [inline csl-data.json]: Made this an "either-or" between `literal` and the rest of the fields.

If no `literal`, then it should at least have `family` (see below)

File: csl-data.json [hunk 11]
```diff
@@ -1,431 +1,440 @@
...
+                                "type": "string"
+                            },
+                            "non-dropping-particle": {
+                                "type": "string"
                             },
-                            "season": {
+                            "suffix": {
+                                "type": "string"
+                            },
+                            "comma-suffix": {
```
Comment (dsifford) [inline csl-data.json]: Not sure what this is, nor `static-ordering` or `parse-names`

File: csl-data.json [hunk 12]
```diff
@@ -1,431 +1,440 @@
...
+            }
+        },
+        "numberDefinition": {
+            "oneOf": [
+                {
+                    "type": "number"
+                },
+                {
+                    "type": "string",
+                    "pattern": "^[0-9]+$"
```
Comment (dsifford) [inline csl-data.json]: Changed this so that if it's a string, it is still just numbers in the string.

**Comment by dsifford** [comment]
Just realized that I removed `literal` from the date definition because I thought they were the same thing.. That's not correct -- `raw` needs to be a parsable date (using ideally the json schema `date-time` [RFC3339](https://tools.ietf.org/html/rfc3339#section-5.6) type), whereas `literal` is a literal date string that isn't parsed or touched in any way.

I'll add that fix in after I hear back.

**Comment by dsifford** [comment]
@rmzelle Yikes... also just realized that there are **dozens** of fields listed here that are not listed in `csl.variables.rnc` --- gotta add those in as well.

File: csl-citation.json (hunk 2 above)
Comment (rmzelle) [reply csl-citation.json]: It's not?

https://github.com/rmzelle/ref-extractor/wiki#single-item-zotero-citation shows an example of how Zotero embeds CSL citation metadata in Word documents. It shows "itemData" as a field of "citationItems", and it contains an array of item metadata as defined by csl-data.json.

File: csl-citation.json (hunk 3 above)
Comment (rmzelle) [reply csl-citation.json]: See https://github.com/rmzelle/ref-extractor/wiki#single-item-zotero-citation again. Zotero and Mendeley both reference the "csl-citation.json" schema in their embedded citation metadata.

File: csl-citation.json (hunk 4 above)
Comment (rmzelle) [reply csl-citation.json]: Any reason why?

I think it was chosen to clarify that this is the ID of the citation object, not of the items referenced within.

File: csl-data.json (hunk 8 above)
Comment (rmzelle) [reply csl-data.json]: It's a toggle to indicate whether a date is uncertain. See http://docs.citationstyles.org/en/stable/specification.html#approximate-dates.

File: csl-data.json (hunk 11 above)
Comment (rmzelle) [reply csl-data.json]: static-ordering: https://github.com/Juris-M/citeproc-js/blob/master/manual/citeproc-doc.rst#non-byzantine-names
parse-names: https://github.com/Juris-M/citeproc-js/blob/master/manual/citeproc-doc.rst#names-1

**Comment by rmzelle** [comment]
> there are ... fields listed in the old csl-data.json that are not listed in csl.variables.rnc --- gotta add those in as well. ... `journalAbbreviation`, `language`, `shortTitle`

None of these have a corresponding CSL variable with the same name.

"language" indicates the language of the item in question, and is used by the CSL processor to determine if special processing is necessary (see e.g. http://docs.citationstyles.org/en/stable/specification.html#non-english-items). We chose not to make it a CSL variable, since it generally wouldn't make much sense to print the field contents in a formatted reference. The field contents should ideally be an easily parseable locale code. And even if the field would contain a human readable language descriptor (e.g. "English"), we don't have the infrastructure in place to translate that for CSL styles set to non-English locales (e.g. to "Anglais" for French).

> What is the difference between `shortTitle` and `title-short`? Can one of these be axed?

Regarding "journalAbbreviation" and "shortTitle", CSL originally only made the abbreviated "short" versions of "container-title" and "title" available in CSL styles through `variable="container-title" form="short"` and `variable="title" form="short"`. The CSL JSON still needed to store these abbreviations somehow, and @fbennett came up with "journalAbbreviation" and "shortTitle". Only later, in CSL 1.0.1, did we allow these abbreviated versions to be called directly through `variable="container-title-short"` and `variable="title-short" form="short"`. At that time, we added corresponding fields to CSL JSON but kept the old fields for backwards compatibility. We should drop these old fields eventually.

See also http://docs.citationstyles.org/en/stable/release-notes.html#new-variables-container-title-short-and-title-short

**Comment by rmzelle** [comment]
@dsifford, do you still have the versions of the CSL JSON schemas on hand right after you upgraded them from schema draft 3 to 7?

I'm happy to review any changes you'd like to suggest to the schema itself, but I much rather have separate pull requests for the schema upgrade and such changes.

File: csl-citation.json (hunk 2 above)
Comment (dsifford) [reply csl-citation.json]: Hm.. Must be a single one-off case built into citeproc-js then. Because that doesn't exist anywhere in my implementation.

File: csl-citation.json (hunk 3 above)
Comment (dsifford) [reply csl-citation.json]: Right, but why cant they use JSON schema as it's indended and just reference the base schema using the top-level `$id` property? It seems redundant and unnecessary.

**Comment by rmzelle** [comment]
> Though a couple will make checks a bit more strict (as noted above).

It would be better to put anything that is not purely related to the draft-3 to draft-7 upgrade in a dedicated pull request.

> After I get around to adding in those 3 missing fields

Add which fields where?

**Comment by dsifford** [comment]
> Add which fields where?

`journalAbbreviation`, `language`, and `shortTitle` are still missing

**Comment by rmzelle** [comment]
> `journalAbbreviation`, `language`, and `shortTitle` are still missing

They're in csl-data.json now:

https://github.com/citation-style-language/schema/blob/f01ba9c5ec2055e381a38598919a379255c496c5/csl-data.json#L63
https://github.com/citation-style-language/schema/blob/f01ba9c5ec2055e381a38598919a379255c496c5/csl-data.json#L66
https://github.com/citation-style-language/schema/blob/f01ba9c5ec2055e381a38598919a379255c496c5/csl-data.json#L69

**Comment by dsifford** [comment]
Right, but I removed them in this PR. I have to add them back in.

**Comment by rmzelle** [comment]
> Right, but I removed them in this PR. I have to add them back in.

Ah, okay. Yes. Sorry for the confusion.

**Comment by dhimmel** [comment]
I created a PR into this PR to update the Python test suite: https://github.com/dsifford/schema/pull/1.

In short, we must upgrade the version of the python `jsonschema` package to support draft-07. Also removed jsonref, because I believe @dsifford fixed the bug that required that package in the first place.

There is still [one failure](https://travis-ci.com/dsifford/schema/builds/95034455#L480) in `test_data_schema_with_extra_property_fails` because it seems extra properties are not causing validation failure. If that is intended, you should change the test.

**Comment by dsifford** [comment]
Nice job @dhimmel... Just pushed 2 of the spec fixes that were talked about above that add back in a few fields I missed. Only thing left to do is rewind some things that aren't fully 100% backwards compatible and then we can probably merge this and open a new PR for draft-7 related schema updates.

File: csl-data.json (hunk 12 above)
Comment (dhimmel) [reply csl-data.json]: I think that will break many things. `edition`, `issue`, `volume`often contain characters. I think it makes sense to save any functional changes for subsequent PRs.

File: csl-data.json [hunk 13]
```diff
@@ -1,431 +1,457 @@
...
-                    "number" 
-                ] 
+            {
+                "$ref": "#/definitions/numberFields"
             },
-            "year-suffix": {
-                "type": "string" 
-            } 
-        },
-        "additionalProperties" : false
```
[the remark this answers is not in the document]
Comment (dsifford) [reply csl-data.json]: Nope, it was not.

Although, just so you're aware, there are some large APIs that currently serve "CSL JSON" responses that are peppered with all kinds of additional properties, so not sure how that would possibly affect those API consumers/maintainers...

<details>
<summary><strong>Here's an example of a typical response from CrossRef</strong></summary>
<pre>
{
    "indexed": {
        "date-parts": [[2018, 12, 18]],
        "date-time": "2018-12-18T19:41:10Z",
        "timestamp": 1545162070737
    },
    "reference-count": 0,
    "publisher": "BMJ",
    "license": [
        {
            "URL": "http://www.bmj.com/company/legal-information/terms-conditions/legal-information/tdm-licencepolicy",
            "start": {
                "date-parts": [[2018, 12, 13]],
                "date-time": "2018-12-13T00:00:00Z",
                "timestamp": 1544659200000
            },
            "delay-in-days": 0,
            "content-version": "tdm"
        }
    ],
    "content-domain": { "domain": ["bmj.com"], "crossmark-restriction": true },
    "abstract": "<jats:title>Abstract</jats:title>\n          <jats:sec>\n            <jats:title>Objective</jats:title>\n            <jats:p>To determine if using a parachute prevents death or major traumatic injury when jumping from an aircraft.</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Design</jats:title>\n            <jats:p>Randomized controlled trial.</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Setting</jats:title>\n            <jats:p>Private or commercial aircraft between September 2017 and August 2018.</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Participants</jats:title>\n            <jats:p>92 aircraft passengers aged 18 and over were screened for participation. 23 agreed to be enrolled and were randomized.</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Intervention</jats:title>\n            <jats:p>Jumping from an aircraft (airplane or helicopter) with a parachute versus an empty backpack (unblinded).</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Main outcome measures</jats:title>\n            <jats:p>Composite of death or major traumatic injury (defined by an Injury Severity Score over 15) upon impact with the ground measured immediately after landing.</jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Results</jats:title>\n            <jats:p>\n              Parachute use did not significantly reduce death or major injury (0% for parachute\n              <jats:italic>v</jats:italic>\n              0% for control; P&gt;0.9). This finding was consistent across multiple subgroups. Compared with individuals screened but not enrolled, participants included in the study were on aircraft at significantly lower altitude (mean of 0.6 m for participants\n              <jats:italic>v</jats:italic>\n              mean of 9146 m for non-participants; P&lt;0.001) and lower velocity (mean of 0 km/h\n              <jats:italic>v</jats:italic>\n              mean of 800 km/h; P&lt;0.001).\n            </jats:p>\n          </jats:sec>\n          <jats:sec>\n            <jats:title>Conclusions</jats:title>\n            <jats:p>Parachute use did not reduce death or major traumatic injury when jumping from aircraft in the first randomized evaluation of this intervention. However, the trial was only able to enroll participants on small stationary aircraft on the ground, suggesting cautious extrapolation to high altitude jumps. When beliefs regarding the effectiveness of an intervention exist in the community, randomized trials might selectively enroll individuals with a lower perceived likelihood of benefit, thus diminishing the applicability of the results to clinical practice.</jats:p>\n          </jats:sec>",
    "DOI": "10.1136/bmj.k5094",
    "type": "article-journal",
    "created": {
        "date-parts": [[2018, 12, 13]],
        "date-time": "2018-12-13T22:34:15Z",
        "timestamp": 1544740455000
    },
    "page": "k5094",
    "update-policy": "http://dx.doi.org/10.1136/crossmarkpolicy",
    "source": "Crossref",
    "is-referenced-by-count": 0,
    "title": "Parachute use to prevent death and major trauma when jumping from aircraft: randomized controlled trial",
    "prefix": "10.1136",
    "author": [
        {
            "given": "Robert W",
            "family": "Yeh",
            "sequence": "first",
            "affiliation": []
        },
        {
            "given": "Linda R",
            "family": "Valsdottir",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Michael W",
            "family": "Yeh",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Changyu",
            "family": "Shen",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Daniel B",
            "family": "Kramer",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Jordan B",
            "family": "Strom",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Eric A",
            "family": "Secemsky",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Joanne L",
            "family": "Healy",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Robert M",
            "family": "Domeier",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Dhruv S",
            "family": "Kazi",
            "sequence": "additional",
            "affiliation": []
        },
        {
            "given": "Brahmajee K",
            "family": "Nallamothu",
            "sequence": "additional",
            "affiliation": []
        }
    ],
    "member": "239",
    "published-online": { "date-parts": [[2018, 12, 13]] },
    "container-title": "BMJ",
    "original-title": [],
    "language": "en",
    "link": [
        {
            "URL": "http://data.bmj.org/tdm/10.1136/bmj.k5094",
            "content-type": "unspecified",
            "content-version": "vor",
            "intended-application": "text-mining"
        },
        {
            "URL": "https://syndication.highwire.org/content/doi/10.1136/bmj.k5094",
            "content-type": "unspecified",
            "content-version": "vor",
            "intended-application": "similarity-checking"
        }
    ],
    "deposited": {
        "date-parts": [[2018, 12, 18]],
        "date-time": "2018-12-18T19:05:10Z",
        "timestamp": 1545159910000
    },
    "score": 1.0,
    "subtitle": [],
    "short-title": [],
    "issued": { "date-parts": [[2018, 12, 13]] },
    "references-count": 0,
    "alternative-id": ["10.1136/bmj.k5094"],
    "URL": "http://dx.doi.org/10.1136/bmj.k5094",
    "relation": {},
    "ISSN": ["0959-8138", "1756-1833"],
    "subject": ["General Medicine"],
    "container-title-short": "BMJ"
}
</pre>
</details>

[the remark this answers is not in the document]
Comment (dhimmel) [reply csl-data.json]: > so not sure how that would possibly affect those API consumers/maintainers...

Not at all... they currently return CSL that does not abide by the JSON Schema. I.e. this is the status quo. For Manubot, we actually use the JSON schema to filter the CSL returned by Crossref and remove all the crap.

**Comment by rmzelle** [comment]
> Only thing left to do is rewind some things that aren't fully 100% backwards compatible and then we can probably merge this and open a new PR for draft-7 related schema updates.

@dsifford, I haven't merged this yet since it's my understanding you still have some rewinding to do.

File: csl-data.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsifford) [reply csl-data.json]: If we continue to use `allOf` and use refs as I do in this PR, then there's no way currently possible to have `additionalProperties: false`. So pick your poison. I personally think it's fine, but if you feel strongly about it, we'd have to scrap this.

cc: @rmzelle

File: csl-data.json (hunk 12 above)
Comment (dsifford) [reply csl-data.json]: To my knowledge, this is the only potentially breaking change that I made. I'll revert this now and then we should be good to go @rmzelle.


@dhimmel feel free to give this another once-over if you'd like.

File: csl-citation.json (hunk 1 above)
Comment (dsifford) [reply csl-citation.json]: This shouldn't be breaking so I'm gonna leave it.

**Comment by dsifford** [comment]
@rmzelle This should be good to go now, pending any objections by @dhimmel

File: csl-data.json [hunk 14]
```diff
@@ -1,431 +1,452 @@
...
+                                    "string",
+                                    "number",
+                                    "boolean"
+                                ]
                             }
                         },
-                        "additionalProperties" : false 
+                        "required": [
+                            "family"
+                        ]
```
Comment (rmzelle) [inline csl-data.json]: I'm not sure we can require "family".

I just checked and a Zotero user can store just a given name, which gives this as exported CSL JSON:

```
			{
				"family": "Nielsen",
				"given": "Jens"
			},
			{
				"family": "",
				"given": "bla"
			}
```

Other implementations might leave out the field entirely.

File: csl-data.json [hunk 15]
```diff
@@ -1,431 +1,452 @@
...
+                },
+                "jurisdiction": {
+                    "type": "string"
+                },
+                "keyword": {
+                    "type": "string"
+                },
+                "language": {
+                    "type": "string"
+                },
```
Comment (rmzelle) [inline csl-data.json]: Maybe it would make sense to carve out a separate definition for non-CSL variable metadata fields like "language"?

File: csl-data.json [hunk 16]
```diff
@@ -1,431 +1,452 @@
...
+                "ISBN": {
+                    "type": "string"
+                },
+                "ISSN": {
+                    "type": "string"
+                },
+                "journalAbbreviation": {
+                    "type": "string"
```
Comment (rmzelle) [inline csl-data.json]: Maybe we should use a separate definition here too, to indicate that "journalAbbreviation" is deprecated in favor of "container-title-short". Idem for "shortTitle" and "title-short".

File: csl-citation.json [hunk 17]
```diff
@@ -1,101 +1,103 @@
...
+    "definitions": {
+        "citationItemDefinition": {
+            "type": "object",
+            "properties": {
+                "id": {
+                    "type": [
+                        "string",
+                        "number"
+                    ]
+                },
```
Comment (rmzelle) [inline csl-citation.json]: The original schema requires "id" for each citation item, which is a requirement we should keep.

File: csl-citation.json [hunk 18]
```diff
@@ -1,101 +1,103 @@
...
+                "itemData": {
+                    "$ref": "csl-data.json"
```
Comment (rmzelle) [inline csl-citation.json]: I haven't tried using your version schema, but are you sure this shouldn't be ` "$ref": "csl-data.json/#/items"` like in the original?

File: csl-data.json [hunk 19]
```diff
@@ -1,431 +1,452 @@
...
-                                        "boolean" 
-                                    ] 
-                                } 
+                                "date-parts": {
+                                    "type": "array",
+                                    "items": {
+                                        "type": "array",
+                                        "items": {
+                                            "type": "number"
+                                        },
```
[the remark this answers is not in the document]
Comment (dhimmel) [reply csl-data.json]: Zotero [currently formats](https://github.com/zotero/zotero/issues/1603) years as strings rather than integers during CSL export.

I think upgrading the schema draft is a great enhancement. However, I think it is crucial that backwards incompatible changes are not made in the initial draft change.

[the remark this answers is not in the document]
Comment (rmzelle) [reply csl-data.json]: There are a few things that got more strict that I didn't flag, like setting required fields for date-parts (either a "date-parts", "literal" or "raw" date), but those seemed pretty basic. Agree we should be careful here.

File: csl-data.json (hunk 11 above)
Comment (fbennett) [reply csl-data.json]: These two attributes were cruft from the early days. I've removed them from fixtures in the test suite, they can be removed from the schema.

File: csl-data.json [hunk 20]
```diff
@@ -1,447 +1,440 @@
...
+                "title-short": {
+                    "type": "string"
+                },
+                "URL": {
+                    "type": "string"
+                },
+                "version": {
+                    "type": "string"
+                },
+                "year-suffix": {
```
Comment (fbennett) [inline csl-data.json]: This is a property of the item in rendered context. Any input value would be ignored.

File: csl-data.json [hunk 21]
```diff
@@ -1,447 +1,440 @@
...
+                "archive-place": {
+                    "type": "string"
+                },
+                "authority": {
+                    "type": "string"
+                },
+                "call-number": {
+                    "type": "string"
+                },
+                "citation-label": {
```
Comment (fbennett) [inline csl-data.json]: This is generated by processors, not set on input.

File: csl-data.json [hunk 22]
```diff
@@ -1,447 +1,440 @@
...
+                    "type": "string"
+                },
+                "citation-number": {
```
Comment (fbennett) [inline csl-data.json]: This is a property of the item in context. Processors generate the value internally. Any input value will be ignored.

File: csl-data.json [hunk 23]
```diff
@@ -1,447 +1,440 @@
...
+                "DOI": {
+                    "type": "string"
+                },
+                "event": {
+                    "type": "string"
+                },
+                "event-place": {
+                    "type": "string"
+                },
+                "first-reference-note-number": {
```
Comment (fbennett) [inline csl-data.json]: This is a processor-generated value that depends on the position of references in a document. Any value set on input will be ignored.

**Comment by dsifford** [comment]
I'll try and look at this again this weekend and address issues raised. Sorry for stalling!

**Comment by rmzelle** [comment]
> I'll try and look at this again this weekend and address issues raised. Sorry for stalling!

@dhimmel, I prepared a minimal update to draft 6 (it looks like no changes are necessary for draft 7) at https://github.com/citation-style-language/schema/pull/168, inspired by this PR. I would prefer to merge that, and then create one or more additional PRs to pull in improvements/changes proposed here (such as using proper definitions etc.).

File: csl-data.json (hunk 21 above)
Comment (dsifford) [reply csl-data.json]: So can this be removed?

File: csl-data.json (hunk 22 above)
Comment (dsifford) [reply csl-data.json]: So can this be removed?

File: csl-data.json (hunk 23 above)
Comment (dsifford) [reply csl-data.json]: So can this be removed?

File: csl-data.json (hunk 20 above)
Comment (dsifford) [reply csl-data.json]: So can this be removed?

**Comment by dsifford** [comment]
@rmzelle Ah, shoot. I missed your message.

I went ahead and addressed all(?) of the comments/concerns raised here. It should theoretically be good to go. But your call.

File: csl-data.json (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsifford) [reply csl-data.json]: Ping @rmzelle -- My comment above still holds relevance.

**Comment by dhimmel** [comment]
Note @dsifford that I'm not a maintainer of this repo, but I appreciate the work you've put in, and hope to see as much of it integrated as makes sense.

Is it possible to split this PR up into smaller ones following #168? I think it will help with review and the pace of things, if the PRs are very focused and do one and only one thing. In addition, each change should indicate whether it functionally changes the schema or not.

> It should theoretically be good to go.

Note that you have a [test failure](https://travis-ci.org/citation-style-language/schema/builds/526713099) of `test_data_schema_with_extra_property_fails`. If you think think this constraint should be removed as per https://github.com/citation-style-language/schema/pull/161#discussion_r247332076, you should either delete that test or add a skip decorator like:

```python
@pytest.mark.skip(reason="Setting additionalProperties to false not possible with allOf")
```

**Comment by clbarnes** [comment]
Any blockers on getting this merged? I'm working on a downstream package and would prefer not to have to start everything over when the updated schema lands.

**Comment by bdarcus** [comment]
Closed this in favor of #168. Any missing pieces can be re-added via dedicated PRs.
